feat: add optional logger wherever possible

ccoVeille opened the pull request:
This commit introduces an optional logger parameter to various structs.
This enhancement allows users to provide custom logging implementations.

This is a naive implementation of #3558

It is provided to iterate on the changes, there are many things that are missing:
- tests
- WithLogger should be added at multiple places

Also, the current discussion in #3558 let me think the current logging interface should be reconsidered, so everything will have to be refactored.

ndyakov commented:
@ccoVeille thank you for opening this PR. I am willing to work with you on improving the logging and do agree that the current interface lacks some useful features. If you have a current need to just be able to set logger per client, we can improve this initially and then continue on the whole logging interface.

ccoVeille commented:
I found a workaround by injecting a dedicated logger in the context and restoring from the context in the current Printf interface.

So I don't have a need right now. We can work together on the target solution you planned initially for the v10 with a fallback for the v9

ndyakov commented:
so @ccoVeille would you consider this pr ready for review ?

ndyakov commented:
@ccoVeille just moved it to Ready for Review and will work on reviewing it this week, thank you.

ndyakov commented on internal/pool/pool.go:
```diff
@@ -961,3 +964,11 @@ func (p *ConnPool) isHealthyConn(cn *Conn, now time.Time) bool {
+	if p.cfg.Logger != nil {
+		logger = *p.cfg.Logger
+	}
+	logger.Printf(ctx, format, args...)
```
```suggestion
	logger := p.cfg.Logger
	if logger == nil {
		logger = internal.Logger
	}
	logger.Printf(ctx, format, args...)
```

I think this shows the precedence more clearly. We could go as far as 
```
	if p.cfg.Logger != nil {
		p.cfg.Logger.Printf(ctx, format, args...)
	} else {
		internal.Logger.Printf(ctx, format, args...)
	}
```
Which would be the more widely used approach.

ccoVeille replied:
I'll update to a pattern like this, here and everywhere.

ndyakov commented on maintnotifications/config.go:
```diff
@@ -128,6 +128,9 @@ type Config struct {
 	MaxHandoffRetries int
+
+	// Logger is an optional custom logger for maintenance notifications.
+	Logger *internal.Logging
```
```suggestion

	// Logger is an optional custom logger for maintenance notifications.
	Logger internal.Logging
```

ndyakov commented on maintnotifications/handoff_worker.go:
```diff
@@ -478,15 +478,23 @@ func (hwm *handoffWorkerManager) closeConnFromRequest(ctx context.Context, reque
+		logger = *hwm.config.Logger
+	}
+	logger.Printf(ctx, format, args...)
+}
```
same as the other `logf` comments

ndyakov commented on options.go:
```diff
@@ -267,6 +268,10 @@ type Options struct {
+
+	// Logger is the logger used by the client for logging.
+	// If none is provided, the global logger [internal.Logger] is used.
+	Logger *internal.Logging
```
this should not be a pointer to an interface.

ndyakov commented on pubsub.go:
```diff
@@ -141,6 +141,17 @@ func mapKeys(m map[string]struct{}) []string {
+	}
+	logger.Printf(ctx, format, args...)
+}
+
```
same as the rest of the `logf` definitions.

ndyakov commented on pubsub.go:
```diff
@@ -631,9 +642,18 @@ func WithChannelSendTimeout(d time.Duration) ChannelOption {
 
+	// Optional logger for logging channel-related messages.
+	Logger *internal.Logging
+
```
those should not be pointers to an interface

ccoVeille commented on internal/pool/pool.go:
```diff
@@ -115,6 +115,9 @@ type Options struct {
 	DialerRetryTimeout time.Duration
+
+	// Optional logger for connection pool operations.
+	Logger *internal.Logging
```
I would have preferred to do not.

I'm unsure why I came to this.

I'll check.

ndyakov commented on maintnotifications/manager.go:
```diff
@@ -318,3 +318,11 @@ func (hm *Manager) AddNotificationHook(notificationHook NotificationHook) {
+		logger = *hm.config.Logger
+	}
+	logger.Printf(ctx, format, args...)
+}
```
There is no rush to do it in one PR. As long as we can keep the solution backwards compatible, we can iterate over the intervals in multiple PRs. Let me know if you need any assistance. You can also reach me on discord.

ccoVeille commented:
@ndyakov  I applied your suggestions to the first commit

Then I have added another commit where I apply something that seems to me to be better.

Let me know what you think.

ccoVeille commented:
I found a panic in the code, I made a fix for this branch, but you may want me to fix it separately

The bug is here, because err can be nil
https://github.com/redis/go-redis/blob/c1766729ec1826a1bfcdcd1d506d4ce44c1734ea/maintnotifications/handoff_worker.go#L484-L490

But here the call to reason.Error() will panic

https://github.com/redis/go-redis/blob/c1766729ec1826a1bfcdcd1d506d4ce44c1734ea/internal/maintnotifications/logs/log_messages.go#L298-L303

Here is the code that spotted it because of my refactoring to always call the logger, and so remove the test on WarnOrAbove
https://github.com/redis/go-redis/actions/runs/19151645751/job/54743116032#step:3:1025

ndyakov reviewed (approved):
Nice approach! Thank you for fixing the panic as well! There is just one comment and I think we can merge this (please sync with master, there is another PR that may interfere with this one, but I will figure out the merging order so we can release next week.

ccoVeille commented on maintnotifications/handoff_worker.go:
```diff
@@ -219,20 +211,17 @@ func (hwm *handoffWorkerManager) processHandoffRequest(request HandoffRequest) {
-				hwm.logf(context.Background(), logs.HandoffFailed(request.ConnID, request.Endpoint, currentRetries, maxRetries, err))
-			}
+				hwm.logger().Infof(context.Background(), logs.HandoffFailed(request.ConnID, request.Endpoint, currentRetries, maxRetries, err))
+
```
it should be better now, please review

ccoVeille commented on internal/pool/pool.go:
```diff
@@ -835,7 +838,7 @@ func (p *ConnPool) removeConn(cn *Conn) {
 		for idx, ic := range p.idleConns {
 			if ic.GetID() == cid {
-				internal.Logger.Printf(context.Background(), "redis: connection pool: removing idle conn[%d]", cid)
+				p.logger().Errorf(context.Background(), "redis: connection pool: removing idle conn[%d]", cid)
```
I'm unsure about the error level here.

Should we decrease to a Warn or an Info ?

ndyakov commented on internal/log.go:
```diff
@@ -77,3 +77,66 @@ func (l LogLevelT) InfoOrAbove() bool {
+	Errorf(ctx context.Context, format string, v ...interface{})
+
+	// Enabled reports whether the given log level is enabled in the logger
+	Enabled(ctx context.Context, level LogLevelT) bool
```
I am not sure about needing a context here even in the future. Can you think of an example in which it will be useful?

ccoVeille replied:
I hesitated. But as I found one on slog.Logger

https://pkg.go.dev/log/slog#Logger.Enabled

ccoVeille replied:
I removed the need to use Enabled in my latest changes

ndyakov commented on internal/maintnotifications/logs/log_messages.go:
```diff
@@ -295,11 +295,10 @@ func RemovingConnectionFromPool(connID uint64, reason error) string {
 	return appendJSONIfDebug(message, map[string]interface{}{
 		"connID": connID,
-		"reason": reason.Error(),
 	})
```
it is fine to have "due to: nil" and "reason": nil
I am still confused how this happened and will have to investigate further once we have this merged.

ccoVeille replied:
@ndyakov please let me know what you want here

ndyakov reviewed (dismissed):
This is starting to look better on each iterations, congrats @ccoVeille.

I have one note / suggestion that we can either implement now and do it in a separate PR:

- Right now we have one official way of setting the log level `SetLogLevel` but at the end this is only needed when there is no custom logger (granted, a user can use the global log level in their own custom logger implementation, but this is not important).

What I would like to suggest is to have a "constructor" function to create a new logger for a given level and then suggest (in the readme and in the documentation) that the users should create custom loggers and pass them in the `options` if they would like to change the log level. Of course, we won't remove the `SetLogLevel` function to be backwards compatible, and it will continue to work, but we can mark it as Deprecated and this can help us transition easily to custom loggers in the future.

ccoVeille commented:
> This is starting to look better on each iterations, congrats @ccoVeille.

You are welcome.

> I have one note / suggestion that we can either implement now and do it in a separate PR:
> 
> - Right now we have one official way of setting the log level `SetLogLevel` but at the end this is only needed when there is no custom logger (granted, a user can use the global log level in their own custom logger implementation, but this is not important).
> 
> What I would like to suggest is to have a "constructor" function to create a new logger for a given level and then suggest (in the readme and in the documentation) that the users should create custom loggers and pass them in the `options` if they would like to change the log level. Of course, we won't remove the `SetLogLevel` function to be backwards compatible, and it will continue to work, but we can mark it as Deprecated and this can help us transition easily to custom loggers in the future. 

We can of course do it in another PR.

But I feel like the current MR should target having code that accepts a logger which signature would match the one of slog.Logger

I feel like your lib should move to structured logger. Not in this PR. Take a look at the hack you have with the metadata you add via json.Encode.

I dislike the idea of merging the PR like this.

I read again slog and log. I have ideas. I will try to find a solution.

ndyakov commented:
@ccoVeille I agree, but don't see a clear way to do it without breaking changes (if we include this PR). Let me think about it a bit. I won't have much time next week, but will get back to it as soon as possible.

ccoVeille commented:
> @ccoVeille I agree, but don't see a clear way to do it without breaking changes (if we include this PR). Let me think about it a bit. I won't have much time next week, but will get back to it as soon as possible.

I refactored everything.

Please take a took. I added structured logging support.

There might be some iterations to validate everything works as expected. I also need to add tests.
But now, it matches what I would expect from the package.

ndyakov commented on maintnotifications/handoff_worker.go:
```diff
@@ -477,16 +463,20 @@ func (hwm *handoffWorkerManager) closeConnFromRequest(ctx context.Context, reque
-		}
-		if internal.LogLevel.WarnOrAbove() {
-			internal.Logger.Printf(ctx, logs.NoPoolProvidedCannotRemove(conn.GetID(), err))
+			hwm.logger().Errorf(ctx, "redis: failed to close connection: %v", err)
```
can you please add a TODO here for me to extract this error as well?

ndyakov reviewed (approved):
overall looks good to me, i will change the messages / calling in the maintnotif package to have more wraps with `Enabled`, but the logger itself looks good. What do you think @ccoVeille, are you happy with this iteration?

ccoVeille commented on logging/legacy.go:
```diff
@@ -0,0 +1,91 @@
+		args = append(args, v[i], v[i+1])
+	}
+
+	return format, args
```
I can adjust. I feel like I can use something like !BADITEM or any marker like sprintf does.

Or simply let sprintf handle it, it would result an error like the one I describe

ccoVeille replied:
Here is the system in the Go code when a slog.Hander is defined as log.SetDefault

https://cs.opensource.google/go/go/+/master:src/log/slog/record.go;l=160-182;drc=8c41a482f9b7a101404cd0b417ac45abd441e598

ccoVeille commented:
But I feel the issue also exists with the filterLogger that introduces another level in the stack when calling Printf

ndyakov reviewed (commented):
overall looks good to me, i can handle the rest of the internal.Logger and do some testing to verify the output is as expected (we do have some e2e tests that assert based on the logs).

ndyakov commented:
@ccoVeille if you don't mind, I will use your branch to update some of the logger calls that are left
